File: searxng-client/searxng_client.py
"""
SearXNG API Client

A Python client for interacting with the SearXNG search API deployed on Railway.
"""

import requests
from typing import List, Dict, Optional
import json


class SearXNGClient:
    """Client for SearXNG search API."""

    def __init__(
        self, base_url: str = "https://searxng-production-b099.up.railway.app"
    ):
        """
        Initialize the SearXNG client.

        Args:
            base_url: The base URL of the SearXNG API
        """
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()

    def search(
        self,
        query: str,
        engines: Optional[List[str]] = None,
        categories: Optional[List[str]] = None,
        pageno: int = 1,
        time_range: Optional[str] = None,
        format: str = "json",
    ) -> Dict:
        """
        Perform a search query.

        Args:
            query: The search query
            engines: List of search engines to use (e.g., ["brave", "duckduckgo", "github"])
            categories: List of categories (e.g., ["general", "images", "news"])
            pageno: Page number for pagination (starts at 1)
            time_range: Time range filter (e.g., "day", "week", "month", "year")
            format: Response format (default: "json")

        Returns:
            Dictionary containing search results

        Example:
            >>> client = SearXNGClient()
            >>> results = client.search("python async", engines=["brave", "github"])
        """
        params = {"q": query, "format": format, "pageno": pageno}

        if engines:
            params["engines"] = ",".join(engines)

        if categories:
            params["categories"] = ",".join(categories)

        if time_range:
            params["time_range"] = time_range

        response = self.session.get(
            f"{self.base_url}/search", params=params, timeout=30
        )
        response.raise_for_status()

        return response.json()

    def search_code(self, query: str) -> Dict:
        """
        Search for code-related results using GitHub and StackOverflow.

        Args:
            query: The search query

        Returns:
            Dictionary containing search results
        """
        return self.search(query, engines=["github", "stackoverflow"])

    def search_academic(self, query: str) -> Dict:
        """
        Search for academic papers using arXiv.

        Args:
            query: The search query

        Returns:
            Dictionary containing search results
        """
        return self.search(query, engines=["arxiv", "semantic scholar"])

    def search_general(self, query: str) -> Dict:
        """
        Search using general search engines (Brave, DuckDuckGo, Startpage).

        Args:
            query: The search query

        Returns:
            Dictionary containing search results
        """
        return self.search(query, engines=["brave", "duckduckgo", "startpage"])

    def get_suggestions(self, query: str) -> List[str]:
        """
        Get search suggestions for a query.

        Args:
            query: The partial search query

        Returns:
            List of suggestion strings
        """
        response = self.session.get(
            f"{self.base_url}/autocompleter", params={"q": query}, timeout=10
        )
        response.raise_for_status()

        data = response.json()
        # Handle both formats: direct list or dict with "suggestions" key
        if isinstance(data, list):
            return data
        return data.get("suggestions", [])


def main():
    """Example usage of the SearXNG client."""
    client = SearXNGClient()

    print("=" * 80)
    print("SearXNG API Client - Example Usage")
    print("=" * 80)

    print("\n1. General Search Example:")
    print("-" * 80)
    results = client.search_general("machine learning frameworks")
    print(f"Query: {results['query']}")
    print(f"Number of results: {len(results['results'])}")
    print("\nTop 3 results:")
    for i, result in enumerate(results["results"][:3], 1):
        print(f"\n{i}. {result['title']}")
        print(f"   URL: {result['url']}")
        print(f"   Engine: {result.get('engine', 'N/A')}")
        if result.get("content"):
            print(f"   Content: {result['content'][:150]}...")

    print("\n" + "=" * 80)
    print("\n2. Code Search Example:")
    print("-" * 80)
    results = client.search_code("python async await")
    print(f"Query: {results['query']}")
    print(f"Number of results: {len(results['results'])}")
    print("\nTop 3 results:")
    for i, result in enumerate(results["results"][:3], 1):
        print(f"\n{i}. {result['title']}")
        print(f"   URL: {result['url']}")
        print(f"   Engine: {result.get('engine', 'N/A')}")

    print("\n" + "=" * 80)
    print("\n3. Academic Search Example:")
    print("-" * 80)
    results = client.search_academic("neural networks attention mechanism")
    print(f"Query: {results['query']}")
    print(f"Number of results: {len(results['results'])}")
    if results["results"]:
        print("\nTop 3 results:")
        for i, result in enumerate(results["results"][:3], 1):
            print(f"\n{i}. {result['title']}")
            print(f"   URL: {result['url']}")
            print(f"   Engine: {result.get('engine', 'N/A')}")

    print("\n" + "=" * 80)
    print("\n4. Search Suggestions Example:")
    print("-" * 80)
    suggestions = client.get_suggestions("python")
    print(f"Suggestions for 'python': {suggestions[:5]}")

    print("\n" + "=" * 80)
    print("\nClient is ready to use!")
    print(f"API Endpoint: {client.base_url}")
    print("=" * 80)


if __name__ == "__main__":
    main()

File: searxng-client/searxng-client.js
/**
 * SearXNG API Client
 * 
 * A JavaScript/TypeScript client for interacting with the SearXNG search API deployed on Railway.
 */

const axios = require('axios');

class SearXNGClient {
    /**
     * Initialize the SearXNG client.
     * 
     * @param {string} baseUrl - The base URL of the SearXNG API
     */
    constructor(baseUrl = 'https://searxng-production-b099.up.railway.app') {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.client = axios.create({
            baseURL: this.baseUrl,
            timeout: 30000,
        });
    }

    /**
     * Perform a search query.
     * 
     * @param {string} query - The search query
     * @param {Object} options - Search options
     * @param {string[]} options.engines - List of search engines to use
     * @param {string[]} options.categories - List of categories
     * @param {number} options.pageno - Page number for pagination (starts at 1)
     * @param {string} options.timeRange - Time range filter
     * @param {string} options.format - Response format (default: "json")
     * @returns {Promise<Object>} Search results
     */
    async search(query, options = {}) {
        const {
            engines = null,
            categories = null,
            pageno = 1,
            timeRange = null,
            format = 'json'
        } = options;

        const params = {
            q: query,
            format,
            pageno
        };

        if (engines) {
            params.engines = engines.join(',');
        }

        if (categories) {
            params.categories = categories.join(',');
        }

        if (timeRange) {
            params.time_range = timeRange;
        }

        const response = await this.client.get('/search', { params });
        return response.data;
    }

    /**
     * Search for code-related results using GitHub and StackOverflow.
     * 
     * @param {string} query - The search query
     * @returns {Promise<Object>} Search results
     */
    async searchCode(query) {
        return this.search(query, {
            engines: ['github', 'stackoverflow']
        });
    }

    /**
     * Search for academic papers using arXiv.
     * 
     * @param {string} query - The search query
     * @returns {Promise<Object>} Search results
     */
    async searchAcademic(query) {
        return this.search(query, {
            engines: ['arxiv', 'semantic scholar']
        });
    }

    /**
     * Search using general search engines (Brave, DuckDuckGo, Startpage).
     * 
     * @param {string} query - The search query
     * @returns {Promise<Object>} Search results
     */
    async searchGeneral(query) {
        return this.search(query, {
            engines: ['brave', 'duckduckgo', 'startpage']
        });
    }

    /**
     * Get search suggestions for a query.
     * 
     * @param {string} query - The partial search query
     * @returns {Promise<string[]>} List of suggestion strings
     */
    async getSuggestions(query) {
        const response = await this.client.get('/autocompleter', {
            params: { q: query }
        });
        const data = response.data;
        // Handle both formats: direct list or dict with "suggestions" key
        if (Array.isArray(data)) {
            return data;
        }
        return data.suggestions || [];
    }
}

/**
 * Example usage of the SearXNG client.
 */
async function main() {
    const client = new SearXNGClient();

    console.log('='.repeat(80));
    console.log('SearXNG API Client - Example Usage');
    console.log('='.repeat(80));

    console.log('\n1. General Search Example:');
    console.log('-'.repeat(80));
    try {
        const results = await client.searchGeneral('machine learning frameworks');
        console.log(`Query: ${results.query}`);
        console.log(`Number of results: ${results.results.length}`);
        console.log('\nTop 3 results:');
        results.results.slice(0, 3).forEach((result, i) => {
            console.log(`\n${i + 1}. ${result.title}`);
            console.log(`   URL: ${result.url}`);
            console.log(`   Engine: ${result.engine || 'N/A'}`);
            if (result.content) {
                console.log(`   Content: ${result.content.substring(0, 150)}...`);
            }
        });
    } catch (error) {
        console.error('Error:', error.message);
    }

    console.log('\n' + '='.repeat(80));
    console.log('\n2. Code Search Example:');
    console.log('-'.repeat(80));
    try {
        const results = await client.searchCode('python async await');
        console.log(`Query: ${results.query}`);
        console.log(`Number of results: ${results.results.length}`);
        console.log('\nTop 3 results:');
        results.results.slice(0, 3).forEach((result, i) => {
            console.log(`\n${i + 1}. ${result.title}`);
            console.log(`   URL: ${result.url}`);
            console.log(`   Engine: ${result.engine || 'N/A'}`);
        });
    } catch (error) {
        console.error('Error:', error.message);
    }

    console.log('\n' + '='.repeat(80));
    console.log('\n3. Academic Search Example:');
    console.log('-'.repeat(80));
    try {
        const results = await client.searchAcademic('neural networks attention mechanism');
        console.log(`Query: ${results.query}`);
        console.log(`Number of results: ${results.results.length}`);
        if (results.results.length > 0) {
            console.log('\nTop 3 results:');
            results.results.slice(0, 3).forEach((result, i) => {
                console.log(`\n${i + 1}. ${result.title}`);
                console.log(`   URL: ${result.url}`);
                console.log(`   Engine: ${result.engine || 'N/A'}`);
            });
        }
    } catch (error) {
        console.error('Error:', error.message);
    }

    console.log('\n' + '='.repeat(80));
    console.log('\n4. Search Suggestions Example:');
    console.log('-'.repeat(80));
    try {
        const suggestions = await client.getSuggestions('python');
        console.log(`Suggestions for 'python': ${suggestions.slice(0, 5).join(', ')}`);
    } catch (error) {
        console.error('Error:', error.message);
    }

    console.log('\n' + '='.repeat(80));
    console.log('\nClient is ready to use!');
    console.log(`API Endpoint: ${client.baseUrl}`);
    console.log('='.repeat(80));
}

// Export for use as module
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { SearXNGClient };
}

// Run examples if executed directly
if (require.main === module) {
    main().catch(console.error);
}
